File: internal/services/balances.go
package services

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"os"
	"strconv"
	"time"

	"github.com/jmoiron/sqlx"
)

const (
	AnnualReturn     = 0.10
	RetirementTarget = 3_500_000.0
	ChartEndYear     = 2069
)

var nonInvestmentTypes = map[string]bool{
	"checking": true, "savings": true, "credit": true, "credit_card": true,
	"loan": true, "mortgage": true, "real_estate": true, "vehicle": true,
	"other_liability": true, "cash": true,
}

func BirthYear() int {
	v, _ := strconv.Atoi(os.Getenv("BIRTH_YEAR"))
	if v == 0 {
		return 1990
	}
	return v
}

func isInvestment(accountType string) bool {
	return !nonInvestmentTypes[accountType] && accountType != ""
}

type ProjectionPoint struct {
	Date  string  `json:"date"`
	Value float64 `json:"value"`
}

type Crossing struct {
	Date string `json:"date"`
	Age  int    `json:"age"`
}

type Projections struct {
	Coast                  []ProjectionPoint `json:"coast"`
	CurrentRate            []ProjectionPoint `json:"current_rate"`
	CoastCrossed           *Crossing         `json:"coast_crossed"`
	ContribCrossed         *Crossing         `json:"contrib_crossed"`
	Contrib401kOnly        []ProjectionPoint `json:"contrib_401k_only"`
	Contrib401kOnlyCrossed *Crossing         `json:"contrib_401k_only_crossed"`
}

// BuildProjections runs monthly compound interest from startDate to endYear.
// Returns coast (no contrib) and currentRate (with monthlyContrib) series.
func BuildProjections(startDate string, startBalance, monthlyContrib float64, endYear int) Projections {
	r := AnnualReturn / 12
	birthYear := BirthYear()

	start, _ := time.Parse("2006-01-02", startDate)
	y, m := start.Year(), int(start.Month())

	coastBal := startBalance
	contribBal := startBalance

	var coastPts, contribPts []ProjectionPoint
	var coastCrossed, contribCrossed *Crossing

	for y <= endYear {
		d := fmt.Sprintf("%04d-%02d-01", y, m)
		coastPts = append(coastPts, ProjectionPoint{Date: d, Value: math.Round(coastBal*100) / 100})
		contribPts = append(contribPts, ProjectionPoint{Date: d, Value: math.Round(contribBal*100) / 100})

		if coastCrossed == nil && coastBal >= RetirementTarget {
			coastCrossed = &Crossing{Date: d[:7], Age: y - birthYear}
		}
		if contribCrossed == nil && contribBal >= RetirementTarget {
			contribCrossed = &Crossing{Date: d[:7], Age: y - birthYear}
		}

		coastBal = coastBal * (1 + r)
		contribBal = contribBal*(1+r) + monthlyContrib

		m++
		if m > 12 {
			m = 1
			y++
		}
	}

	return Projections{
		Coast:          coastPts,
		CurrentRate:    contribPts,
		CoastCrossed:   coastCrossed,
		ContribCrossed: contribCrossed,
	}
}

type BalancePoint struct {
	Date  string  `db:"date"  json:"date"`
	Value float64 `db:"total" json:"value"`
}

// GetBalanceHistory returns summed brokerage balances by date (ascending),
// excluding accounts named "HSA Cash".
func GetBalanceHistory(db *sqlx.DB) ([]BalancePoint, error) {
	var ids []int
	if err := db.Select(&ids, `SELECT a.id FROM accounts a
		WHERE a.type = 'brokerage' AND a.name != 'HSA Cash'`); err != nil || len(ids) == 0 {
		return nil, err
	}

	query, args, _ := sqlx.In(`
		SELECT ab.date, ROUND(SUM(ab.balance),2) as total
		FROM account_balances ab WHERE ab.account_id IN (?)
		GROUP BY ab.date ORDER BY ab.date ASC`, ids)
	query = db.Rebind(query)

	var pts []BalancePoint
	return pts, db.Select(&pts, query, args...)
}

type MonthlyContrib struct {
	Total       float64
	Monthly401k float64
	MonthlyRoth float64
	AvgBiweekly float64
}

func GetMonthlyContrib(db *sqlx.DB) (MonthlyContrib, error) {
	var avg float64
	err := db.QueryRow(`SELECT COALESCE(AVG(total_401k),0) FROM (
		SELECT total_401k FROM pay_statements
		WHERE total_401k > 200 ORDER BY rowid DESC LIMIT 24
	)`).Scan(&avg)
	if err != nil {
		return MonthlyContrib{}, err
	}
	monthly401k := avg * 26 / 12
	monthlyRoth := (7500.0 * 2) / 12
	return MonthlyContrib{
		Total:       monthly401k + monthlyRoth,
		Monthly401k: monthly401k,
		MonthlyRoth: monthlyRoth,
		AvgBiweekly: avg,
	}, nil
}

type BalanceChartData struct {
	TotalHistory        []BalancePoint `json:"total_history"`
	Projections         Projections    `json:"projections"`
	MonthlyContribution float64        `json:"monthly_contribution"`
	Monthly401k         float64        `json:"monthly_401k"`
	MonthlyRoth         float64        `json:"monthly_roth"`
	AvgBiweekly401k     float64        `json:"avg_biweekly_401k"`
	ActiveBalance       float64        `json:"active_balance"`
	AnnualReturn        float64        `json:"annual_return"`
	BirthYear           int            `json:"birth_year"`
	Target              float64        `json:"target"`
	ChartStart          string         `json:"chart_start"`
}

func GetBalanceChartData(db *sqlx.DB) (BalanceChartData, error) {
	mc, err := GetMonthlyContrib(db)
	if err != nil {
		return BalanceChartData{}, err
	}

	history, err := GetBalanceHistory(db)
	if err != nil {
		return BalanceChartData{}, err
	}

	var activeBalance float64
	db.QueryRow(`SELECT COALESCE(SUM(ab.balance),0) FROM account_balances ab
		JOIN accounts a ON ab.account_id = a.id
		WHERE a.type = 'brokerage' AND a.name != 'HSA Cash'
		  AND ab.balance > 0
		  AND ab.date = (SELECT MAX(date) FROM account_balances WHERE account_id = ab.account_id)
	`).Scan(&activeBalance)

	var proj Projections
	if len(history) > 0 {
		last := history[len(history)-1]
		proj = BuildProjections(last.Date, activeBalance, mc.Total, ChartEndYear)
		p401k := BuildProjections(last.Date, activeBalance, mc.Monthly401k, ChartEndYear)
		proj.Contrib401kOnly = p401k.CurrentRate
		proj.Contrib401kOnlyCrossed = p401k.ContribCrossed
	}

	// Chart start date from oldest pay statement
	chartStart := "2022-04-01"
	var rawDate string
	if err := db.QueryRow("SELECT pay_date FROM pay_statements ORDER BY rowid ASC LIMIT 1").Scan(&rawDate); err == nil {
		iso := ToISO(rawDate)
		if len(iso) >= 7 {
			chartStart = iso[:7] + "-01"
		}
	}

	return BalanceChartData{
		TotalHistory:        history,
		Projections:         proj,
		MonthlyContribution: mc.Total,
		Monthly401k:         mc.Monthly401k,
		MonthlyRoth:         mc.MonthlyRoth,
		AvgBiweekly401k:     mc.AvgBiweekly,
		ActiveBalance:       activeBalance,
		AnnualReturn:        AnnualReturn,
		BirthYear:           BirthYear(),
		Target:              RetirementTarget,
		ChartStart:          chartStart,
	}, nil
}

// SyncResult holds counts from a Monarch sync.
type SyncResult struct {
	Synced    int
	Snapshots int
}

// MonarchSync fetches current balances and historical snapshots from Monarch Money.
func MonarchSync(db *sqlx.DB) (SyncResult, error) {
	sessionID := os.Getenv("MONARCH_SESSION_ID")
	csrfToken := os.Getenv("MONARCH_CSRF_TOKEN")
	cfClearance := os.Getenv("MONARCH_CF_CLEARANCE")

	// Config table overrides env
	rows, _ := db.Query(`SELECT key, value FROM config WHERE key IN (
		'monarch_session_id','monarch_csrf_token','monarch_cf_clearance')`)
	if rows != nil {
		defer rows.Close()
		for rows.Next() {
			var k, v string
			rows.Scan(&k, &v)
			switch k {
			case "monarch_session_id":
				sessionID = v
			case "monarch_csrf_token":
				csrfToken = v
			case "monarch_cf_clearance":
				cfClearance = v
			}
		}
	}

	if sessionID == "" || csrfToken == "" {
		return SyncResult{}, fmt.Errorf("MONARCH_SESSION_ID and MONARCH_CSRF_TOKEN must be set")
	}

	gql := func(query string, variables map[string]interface{}) (map[string]interface{}, error) {
		body, _ := json.Marshal(map[string]interface{}{"query": query, "variables": variables})
		req, _ := http.NewRequest("POST", "https://api.monarch.com/graphql", bytes.NewReader(body))
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Cookie", fmt.Sprintf("cf_clearance=%s; csrftoken=%s; session_id=%s", cfClearance, csrfToken, sessionID))
		req.Header.Set("x-csrftoken", csrfToken)
		req.Header.Set("client-platform", "web")
		req.Header.Set("Origin", "https://app.monarch.com")
		req.Header.Set("User-Agent", "Mozilla/5.0 (X11; Linux x86_64; rv:150.0) Gecko/20100101 Firefox/150.0")

		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			return nil, err
		}
		defer resp.Body.Close()
		if resp.StatusCode == 401 || resp.StatusCode == 403 {
			return nil, fmt.Errorf("Monarch session expired — run monarch_reauth.py")
		}
		var out map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&out)
		return out, nil
	}

	result, err := gql(`{ accounts { id displayName currentBalance type { name } institution { name } } }`, nil)
	if err != nil {
		return SyncResult{}, err
	}

	data, _ := result["data"].(map[string]interface{})
	accounts, _ := data["accounts"].([]interface{})
	today := time.Now().Format("2006-01-02")
	res := SyncResult{}

	for _, a := range accounts {
		acc, ok := a.(map[string]interface{})
		if !ok {
			continue
		}
		monarchID, _ := acc["id"].(string)
		name, _ := acc["displayName"].(string)
		accType := ""
		if t, ok := acc["type"].(map[string]interface{}); ok {
			accType, _ = t["name"].(string)
		}
		institution := ""
		if inst, ok := acc["institution"].(map[string]interface{}); ok {
			institution, _ = inst["name"].(string)
		}
		balance, _ := acc["currentBalance"].(float64)

		db.Exec(`INSERT INTO accounts (monarch_id, name, type, institution) VALUES (?,?,?,?)
			ON CONFLICT(monarch_id) DO UPDATE SET name=excluded.name, type=excluded.type, institution=excluded.institution`,
			monarchID, name, accType, institution)

		var accountID int
		db.QueryRow("SELECT id FROM accounts WHERE monarch_id=?", monarchID).Scan(&accountID)

		db.Exec(`INSERT OR REPLACE INTO account_balances (account_id, date, balance) VALUES (?,?,?)`,
			accountID, today, balance)
		res.Synced++

		if isInvestment(accType) {
			time.Sleep(300 * time.Millisecond)
			hist, err := gql(`query($id: UUID!) {
				snapshots: snapshotsForAccount(accountId: $id) { date signedBalance }
			}`, map[string]interface{}{"id": monarchID})
			if err != nil {
				continue
			}
			hdata, _ := hist["data"].(map[string]interface{})
			snaps, _ := hdata["snapshots"].([]interface{})
			for _, s := range snaps {
				snap, ok := s.(map[string]interface{})
				if !ok {
					continue
				}
				d, _ := snap["date"].(string)
				if len(d) > 10 {
					d = d[:10]
				}
				v, _ := snap["signedBalance"].(float64)
				if d != "" {
					db.Exec(`INSERT OR REPLACE INTO account_balances (account_id, date, balance) VALUES (?,?,?)`,
						accountID, d, v)
					res.Snapshots++
				}
			}
		}
	}
	return res, nil
}

File: app/routes/balances.py
import json
import os
import urllib.error
import urllib.request
from datetime import date

from flask import Blueprint, render_template, jsonify, request
from app.db import get_db

bp = Blueprint("balances", __name__)

_GQL_URL = "https://api.monarch.com/graphql"


def _monarch_credentials():
    """DB config takes precedence over .env for session tokens."""
    db = get_db()
    rows = {r["key"]: r["value"] for r in db.execute(
        "SELECT key, value FROM config WHERE key IN (?,?,?)",
        ("monarch_session_id", "monarch_csrf_token", "monarch_cf_clearance"),
    ).fetchall()}
    return (
        rows.get("monarch_session_id")   or os.environ.get("MONARCH_SESSION_ID", ""),
        rows.get("monarch_csrf_token")   or os.environ.get("MONARCH_CSRF_TOKEN", ""),
        rows.get("monarch_cf_clearance") or os.environ.get("MONARCH_CF_CLEARANCE", ""),
    )


def _monarch_gql(query, variables=None):
    session_id, csrf_token, cf_clearance = _monarch_credentials()
    if not session_id or not csrf_token:
        raise RuntimeError("MONARCH_SESSION_ID and MONARCH_CSRF_TOKEN must be set in .env")
    cookies = f"cf_clearance={cf_clearance}; csrftoken={csrf_token}; session_id={session_id}"
    body = json.dumps({"query": query, "variables": variables or {}}).encode()
    req = urllib.request.Request(_GQL_URL, data=body, headers={
        "Content-Type": "application/json",
        "Cookie": cookies,
        "x-csrftoken": csrf_token,
        "client-platform": "web",
        "Origin": "https://app.monarch.com",
        "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:150.0) Gecko/20100101 Firefox/150.0",
    })
    try:
        resp = urllib.request.urlopen(req)
        return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        if e.code in (401, 403):
            raise RuntimeError(
                "Monarch session expired — run: python3 monarch_reauth.py"
            )
        raise


def _forward_fill_totals(rows):
    """Sum account balances by date, carrying each account's last known value forward.

    Prevents dates where only one account has a snapshot (e.g. Coinbase daily)
    from dragging the total down while others are silent.
    """
    from collections import defaultdict
    acct_history = defaultdict(list)
    for r in rows:
        acct_history[r["id"]].append((r["date"], r["balance"] or 0))

    all_dates = sorted({r["date"] for r in rows})
    latest = {}
    total_by_date = {}
    for d in all_dates:
        for aid, history in acct_history.items():
            for (date, bal) in history:
                if date == d:
                    latest[aid] = bal
        total_by_date[d] = sum(latest.values())
    return total_by_date


def _balance_rows(db, account_id=None):
    if account_id:
        return db.execute("""
            SELECT ab.date, a.name, a.type, a.institution, ab.balance
            FROM account_balances ab
            JOIN accounts a ON ab.account_id = a.id
            WHERE a.id = ?
            ORDER BY ab.date DESC, a.name ASC
        """, (account_id,)).fetchall()
    return db.execute("""
        SELECT ab.date, a.name, a.type, a.institution, ab.balance
        FROM account_balances ab
        JOIN accounts a ON ab.account_id = a.id
        ORDER BY ab.date DESC, a.name ASC
    """).fetchall()


@bp.route("/balances")
def balances():
    db = get_db()
    accounts = db.execute("SELECT id, name FROM accounts ORDER BY name").fetchall()
    rows = _balance_rows(db, request.args.get("account_id"))
    return render_template("balances.html", accounts=accounts, rows=rows)


@bp.route("/balances-table")
def balances_table():
    db = get_db()
    rows = _balance_rows(db, request.args.get("account_id"))
    return render_template("partials/balances_table.html", rows=rows)


@bp.route("/api/monarch/sync", methods=["POST"])
def monarch_sync():
    import time

    try:
        result = _monarch_gql("""{ accounts {
            id displayName currentBalance
            type { name }
            institution { name }
        } }""")
    except RuntimeError as e:
        return jsonify({"error": str(e)}), 500
    except Exception as e:
        return jsonify({"error": f"Monarch request failed: {e}"}), 500

    accounts = result.get("data", {}).get("accounts", [])
    db = get_db()
    today = date.today().isoformat()
    synced_accounts = 0
    synced_snapshots = 0

    for acc in accounts:
        monarch_id = acc.get("id")
        name = acc.get("displayName") or ""
        acc_type = (acc.get("type") or {}).get("name", "")
        institution = (acc.get("institution") or {}).get("name", "")
        balance = acc.get("currentBalance")
        if monarch_id is None:
            continue

        db.execute(
            "INSERT INTO accounts (monarch_id, name, type, institution) VALUES (?,?,?,?) "
            "ON CONFLICT(monarch_id) DO UPDATE SET name=excluded.name, type=excluded.type, institution=excluded.institution",
            (monarch_id, name, acc_type, institution),
        )
        db.commit()
        account_id = db.execute(
            "SELECT id FROM accounts WHERE monarch_id=?", (monarch_id,)
        ).fetchone()["id"]

        if balance is not None:
            db.execute(
                "INSERT OR REPLACE INTO account_balances (account_id, date, balance) VALUES (?,?,?)",
                (account_id, today, balance),
            )
            db.commit()
            synced_accounts += 1

        # Pull full snapshot history to fill gaps — skip if synced within 7 days
        if _is_investment(acc_type):
            recent = db.execute("""
                SELECT MAX(date) as last FROM account_balances WHERE account_id=?
            """, (account_id,)).fetchone()["last"]
            from datetime import timedelta
            cutoff = (date.today() - timedelta(days=7)).isoformat()
            if recent and recent >= cutoff:
                continue
            try:
                hist = _monarch_gql("""
                    query($id: UUID!) {
                        snapshots: snapshotsForAccount(accountId: $id) {
                            date
                            signedBalance
                        }
                    }
                """, {"id": monarch_id})
                snaps = hist.get("data", {}).get("snapshots", []) or []
                for s in snaps:
                    d = (s.get("date") or "")[:10]
                    v = s.get("signedBalance")
                    if d and v is not None:
                        db.execute(
                            "INSERT OR REPLACE INTO account_balances (account_id, date, balance) VALUES (?,?,?)",
                            (account_id, d, v),
                        )
                        synced_snapshots += 1
                db.commit()
                time.sleep(0.3)
            except Exception:
                pass

    return jsonify({"synced": synced_accounts, "snapshots": synced_snapshots})


_NON_INVESTMENT_TYPES = {
    "checking", "savings", "credit", "credit_card", "loan",
    "mortgage", "real_estate", "vehicle", "other_liability", "cash",
}

BIRTH_YEAR = int(os.environ.get("BIRTH_YEAR", 1990))
RETIREMENT_TARGET = 3_500_000
ANNUAL_RETURN = 0.10
CHART_END_YEAR = 2069  # age 75


def _is_investment(account_type: str) -> bool:
    t = (account_type or "").lower().strip()
    return t not in _NON_INVESTMENT_TYPES and t != ""


def _build_projections(start_date_str: str, start_balance: float, monthly_contribution: float):
    from datetime import date as date_cls
    import calendar

    r = ANNUAL_RETURN / 12
    start = date_cls.fromisoformat(start_date_str)

    coast_points, contrib_points = [], []
    coast_crossed = contrib_crossed = None

    coast_bal = contrib_bal = start_balance
    y, m = start.year, start.month

    while y <= CHART_END_YEAR:
        d = f"{y:04d}-{m:02d}-01"
        coast_points.append({"date": d, "value": round(coast_bal, 2)})
        contrib_points.append({"date": d, "value": round(contrib_bal, 2)})

        if coast_crossed is None and coast_bal >= RETIREMENT_TARGET:
            coast_crossed = {"date": d, "age": y - BIRTH_YEAR}
        if contrib_crossed is None and contrib_bal >= RETIREMENT_TARGET:
            contrib_crossed = {"date": d, "age": y - BIRTH_YEAR}

        coast_bal = coast_bal * (1 + r)
        contrib_bal = contrib_bal * (1 + r) + monthly_contribution

        m += 1
        if m > 12:
            m = 1
            y += 1

    return coast_points, contrib_points, coast_crossed, contrib_crossed


@bp.route("/api/balance-chart-data")
def balance_chart_data():
    db = get_db()

    EXCLUDE = {'HSA Cash'}

    # All brokerage accounts for historical totals (includes Investment History etc.)
    all_brokerage_ids = {
        r["id"]
        for r in db.execute("""
            SELECT a.id FROM accounts a WHERE a.type = 'brokerage' AND a.name NOT IN ({})
        """.format(",".join("?" * len(EXCLUDE))), list(EXCLUDE)).fetchall()
    }

    # Only accounts with positive current balance for projection starting point
    active_ids = {
        r["account_id"]
        for r in db.execute("""
            SELECT ab.account_id
            FROM account_balances ab
            JOIN accounts a ON ab.account_id = a.id
            WHERE a.type = 'brokerage'
              AND a.name NOT IN ({})
              AND ab.date = (SELECT MAX(date) FROM account_balances WHERE account_id = ab.account_id)
              AND ab.balance > 0
        """.format(",".join("?" * len(EXCLUDE))), list(EXCLUDE)).fetchall()
    }

    if not all_brokerage_ids:
        return jsonify({"total_history": [], "projections": {}, "monthly_contribution": 0, "target": RETIREMENT_TARGET})

    placeholders = ",".join("?" * len(all_brokerage_ids))
    rows = db.execute(f"""
        SELECT ab.date, a.id, ab.balance
        FROM account_balances ab
        JOIN accounts a ON ab.account_id = a.id
        WHERE a.id IN ({placeholders})
        ORDER BY ab.date ASC
    """, list(all_brokerage_ids)).fetchall()

    total_by_date = _forward_fill_totals(rows)
    sorted_dates = sorted(total_by_date)
    total_history = [{"date": d, "value": round(total_by_date[d], 2)} for d in sorted_dates]

    # Average 401k from last 24 regular paychecks (excludes tiny supplemental entries)
    contrib_row = db.execute("""
        SELECT AVG(total_401k) as avg FROM (
            SELECT total_401k FROM pay_statements
            WHERE total_401k > 200
            ORDER BY rowid DESC LIMIT 24
        )
    """).fetchone()
    avg_biweekly_401k = float(contrib_row["avg"] or 0)
    monthly_401k = avg_biweekly_401k * 26 / 12

    # Max Roth IRA for 2 people ($7,500/person/year as of 2026)
    monthly_roth = (7_500 * 2) / 12

    monthly_contribution = monthly_401k + monthly_roth

    # Projection starts from current active balance only (not zeroed-out historical accounts)
    active_current = db.execute(f"""
        SELECT SUM(ab.balance) as total
        FROM account_balances ab
        WHERE ab.account_id IN ({",".join("?" * len(active_ids))})
          AND ab.date = (SELECT MAX(date) FROM account_balances WHERE account_id = ab.account_id)
    """, list(active_ids)).fetchone()["total"] or 0

    projections = {}
    if total_history:
        last = total_history[-1]
        coast, contrib, coast_x, contrib_x = _build_projections(
            last["date"], round(active_current, 2), monthly_contribution
        )
        _, contrib_401k, _, contrib_401k_x = _build_projections(
            last["date"], round(active_current, 2), monthly_401k
        )
        projections = {
            "coast": coast,
            "current_rate": contrib,
            "coast_crossed": coast_x,
            "contrib_crossed": contrib_x,
            "contrib_401k_only": contrib_401k,
            "contrib_401k_only_crossed": contrib_401k_x,
        }

    start_row = db.execute("""
        SELECT pay_date FROM pay_statements ORDER BY rowid ASC LIMIT 1
    """).fetchone()
    raw = start_row["pay_date"] if start_row else "4/1/2022"
    parts = raw.split("/")
    chart_start = f"{int(parts[2]):04d}-{int(parts[0]):02d}-01"

    return jsonify({
        "total_history": total_history,
        "projections": projections,
        "monthly_contribution": round(monthly_contribution, 2),
        "monthly_401k": round(monthly_401k, 2),
        "monthly_roth": round(monthly_roth, 2),
        "avg_biweekly_401k": round(avg_biweekly_401k, 2),
        "active_balance": round(active_current, 2),
        "annual_return": ANNUAL_RETURN,
        "birth_year": BIRTH_YEAR,
        "target": RETIREMENT_TARGET,
        "chart_start": chart_start,
    })
